From 892d0ed91adebc66df0d0dc9884d71f276dce564 Mon Sep 17 00:00:00 2001
From: Jan Kara <jack@suse.cz>
Date: Thu, 22 Jun 2017 09:12:58 +0200
Subject: [PATCH 08/11] ocfs2: Don't clear SGID when inheriting ACLs
Patch-mainline: v4.13-rc4
Git-commit: 19ec8e48582670c021e998b9deb88e39a842ff45
References: bsc#1030552

When new directory 'DIR1' is created in a directory 'DIR0' with SGID bit
set, DIR1 is expected to have SGID bit set (and owning group equal to
the owning group of 'DIR0'). However when 'DIR0' also has some default
ACLs that 'DIR1' inherits, setting these ACLs will result in SGID bit on
'DIR1' to get cleared if user is not member of the owning group.

Fix the problem by moving posix_acl_update_mode() out of ocfs2_set_acl()
into ocfs2_iop_set_acl(). That way the function will not be called when
inheriting ACLs which is what we want as it prevents SGID bit clearing
and the mode has been properly set by posix_acl_create() anyway. Also
posix_acl_chmod() that is calling ocfs2_set_acl() takes care of updating
mode itself.

Fixes: 073931017b49d9458aa351605b43a7e34598caef
Cc: stable@vger.kernel.org
Cc: Joel Becker <jlbec@evilplan.org>
Cc: ocfs2-devel@oss.oracle.com
Signed-off-by: Jan Kara <jack@suse.cz>

---
 fs/ocfs2/acl.c |   19 ++++++++++---------
 1 file changed, 10 insertions(+), 9 deletions(-)

--- a/fs/ocfs2/acl.c
+++ b/fs/ocfs2/acl.c
@@ -246,15 +246,6 @@ static int ocfs2_set_acl(handle_t *handl
 	switch (type) {
 	case ACL_TYPE_ACCESS:
 		name_index = OCFS2_XATTR_INDEX_POSIX_ACL_ACCESS;
-		if (acl) {
-			umode_t mode;
-			ret = posix_acl_update_mode(inode, &mode, &acl);
-			if (ret)
-				return ret;
-			ret = ocfs2_acl_set_mode(inode, di_bh, handle, mode);
-			if (ret)
-				return ret;
-		}
 		break;
 	case ACL_TYPE_DEFAULT:
 		name_index = OCFS2_XATTR_INDEX_POSIX_ACL_DEFAULT;
@@ -497,9 +488,19 @@ static int ocfs2_xattr_set_acl(struct de
 		if (IS_ERR(acl))
 			return PTR_ERR(acl);
 		else if (acl) {
+			umode_t mode;
+
 			ret = posix_acl_valid(acl);
 			if (ret)
 				goto cleanup;
+
+			ret = posix_acl_update_mode(inode, &mode, &acl);
+			if (ret)
+				goto cleanup;
+
+			ret = ocfs2_acl_set_mode(inode, NULL, NULL, mode);
+			if (ret)
+				goto cleanup;
 		}
 	} else
 		acl = NULL;
